pkg/recommender/recommender.go

```diff
@@ -0,0 +1,129 @@
+package recommender
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/pkg/errors"
+	"github.com/scylladb/go-log"
+	"github.com/scylladb/scylla-operator-autoscaler/pkg/api/v1alpha1"
+	"github.com/scylladb/scylla-operator-autoscaler/pkg/recommender/metrics"
+	scyllav1 "github.com/scylladb/scylla-operator/pkg/api/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type (
+	Recommender interface {
+		RunOnce(ctx context.Context) error
+	}
+
+	recommender struct {
+		client          client.Client
+		logger          log.Logger
+		metricsProvider metrics.Provider
+	}
+)
+
+func New(ctx context.Context, c client.Client, logger log.Logger, metricsSelector map[string]string) (Recommender, error) {
+	p, err := metrics.NewPrometheusProvider(ctx, c, logger, metricsSelector)
+	if err != nil {
+		return nil, errors.Wrap(err, "create metrics provider")
+	}
+
+	return &recommender{
+		client:          c,
+		logger:          logger,
+		metricsProvider: p,
+	}, nil
+}
+
+func (r *recommender) RunOnce(ctx context.Context) error {
+	scas := &v1alpha1.ScyllaClusterAutoscalerList{}
+	err := r.client.List(ctx, scas)
+	if err != nil {
+		return errors.Wrap(err, "fetch SCAs")
+	}
+
+	for idx, _ := range scas.Items {
+		sca := &scas.Items[idx]
+		targetRef := sca.Spec.TargetRef
+
+		sc := &scyllav1.ScyllaCluster{}
+		err = r.client.Get(ctx, client.ObjectKey{
+			Namespace: targetRef.Namespace,
+			Name:      targetRef.Name,
+		}, sc)
+		if err != nil {
+			r.logger.Error(ctx, "fetch referenced ScyllaCluster", "error", err)
+			continue
+		}
+
+		ruleGroups := sca.Spec.ScalingRuleGroups
+		if len(sca.Spec.ScalingPolicy.Datacenters) == 0 {
+			continue
+		}
+
+		dc := sc.Spec.Datacenter
+		rackRecommendations := make([]v1alpha1.RackRecommendations, 0)
+		for _, rackScalingPolicy := range sca.Spec.ScalingPolicy.Datacenters[0].RackScalingPolicies {
+			rackName := rackScalingPolicy.Name
+			i := 0
+			for i := range dc.Racks {
+				if dc.Racks[i].Name == rackName {
+					break
+				}
+			}
+			if i == len(dc.Racks) {
+				r.logger.Error(ctx, "find rack", "error", errors.New("rack not found"), "rack", rackName)
+				continue
+			}
```

Comment:
instead keeping iterator, you can define boolean variable indicating wheter value was found or not.